src/GraphQL/Types/Schema.cs

```diff
@@ -371,16 +371,36 @@ IEnumerable<ISchemaNodeVisitor> GetVisitors()
                 }
             }
 
-            _allTypes = new SchemaTypes(this, _services);
+            _allTypes = CreateSchemaTypes();
 
-            // At this point, Initialized will return false, and Initialize will still lock while waiting for initialization to complete.
-            // However, AllTypes and similar properties will return a reference to SchemaTypes without waiting for a lock.
-            _allTypes.ApplyMiddleware(FieldMiddleware);
+            try
+            {
+                // At this point, Initialized will return false, and Initialize will still lock while waiting for initialization to complete.
+                // However, AllTypes and similar properties will return a reference to SchemaTypes without waiting for a lock.
+                _allTypes.ApplyMiddleware(FieldMiddleware);
 
-            foreach (var visitor in GetVisitors())
-                visitor.Run(this);
+                foreach (var visitor in GetVisitors())
+                    visitor.Run(this);
 
-            Validate();
+                Validate();
+            }
+            catch
+            {
+                _allTypes = null;
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Creates and returns a new instance of <see cref="SchemaTypes"/> for this schema.
+        /// Does not apply middleware, apply schema visitors, or validate the schema.
+        /// </summary>
+        /// <remarks>
+        /// This executes within a lock in <see cref="Initialize"/>.
+        /// </remarks>
+        protected virtual SchemaTypes CreateSchemaTypes()
+        {
+            return new SchemaTypes(this, _services);
```

Comment:
Overridden method still can not get _services.

Reply:
True but if they inherit from Schema they can capture the service provider in their own constructor